src/cargo_communication/cargo_messages_handle.rs

```diff
@@ -0,0 +1,258 @@
+    }
+
+    fn handle_information_from_test(&mut self, test_type: TestType) {
+        match test_type {
+            TestType::Suite(event) => self.handle_test_suite(event),
+            TestType::Test(event) => self.handle_single_test(event),
+        }
+    }
+
+    fn handle_test_suite(&mut self, event: SuiteEvent) {
+        if let TaskState::Test(test_state) = &mut self.state.task_state {
+            let mut task_id = test_state.suite_test_task_id.clone();
+            match event {
+                SuiteEvent::Started(s) => {
+                    let new_id = generate_random_id();
+                    task_id.id = new_id.clone();
+                    test_state.suite_test_task_id.id = new_id;
+                    test_state.suite_task_progress.total = s.test_count as i64;
+                    self.report_task_start(
+                        task_id,
+                        None,
+                        // TODO change target to actual BuildTargetIdentifier
+                        Some(TaskDataWithKind::TestTask(TestTaskData {
+                            target: Default::default(),
+                        })),
+                    );
+                }
+                SuiteEvent::Ok(result) => self.report_task_finish(
+                    task_id,
+                    StatusCode::Ok,
+                    None,
+                    Some(result.to_test_report()),
+                ),
+                SuiteEvent::Failed(result) => self.report_task_finish(
+                    task_id,
+                    StatusCode::Error,
+                    None,
+                    Some(result.to_test_report()),
+                ),
+            }
+        }
+    }
+
+    fn handle_single_test(&mut self, event: TestEvent) {
+        if let TaskState::Test(test_state) = &mut self.state.task_state {
+            match event {
+                TestEvent::Started(started) => {
+                    let test_task_id = generate_task_id(&test_state.suite_test_task_id);
+                    test_state
+                        .single_test_task_ids
+                        .insert(started.name.clone(), test_task_id.clone());
+                    self.report_task_start(
+                        test_task_id,
+                        None,
+                        Some(TaskDataWithKind::TestStart(TestStartData {
+                            display_name: started.name,
+                            // TODO add location of build target
+                            location: None,
+                        })),
+                    );
+                }
+                TestEvent::Ok(result) => self.finish_single_test(result, TestStatus::Passed),
+                TestEvent::Failed(result) => self.finish_single_test(result, TestStatus::Failed),
+                TestEvent::Ignored(result) => self.finish_single_test(result, TestStatus::Ignored),
+                TestEvent::Timeout(result) => self.finish_single_test(result, TestStatus::Failed),
+            }
```

Comment:
można by zaimplementować `impl From<TestEvent> for TestStatus` i zamiast tych 4 branch zrobić jeden coś jak
```
TestEvent::Ok(result) | TestEvent::Failed(result) | ... => self.finish_single_test(result, event.into()),
```

Reply:
Yet another problem przy wydzielaniu do crate'a
Można by ewentualnie zrobić funkcję dla TestEvent zwracającą TestStatus, jeżeli byłoby to lepsze niż aktualne rozwiązanie

Reply:
imo można tak zrobić